go/engine/login_offline.go

```diff
@@ -0,0 +1,181 @@
+package engine
+
+import (
+	"errors"
+
+	"github.com/keybase/client/go/libkb"
+	"github.com/keybase/client/go/protocol/keybase1"
+)
+
+type LoginOffline struct {
+	libkb.Contextified
+}
+
+func NewLoginOffline(g *libkb.GlobalContext) *LoginOffline {
+	return &LoginOffline{
+		Contextified: libkb.NewContextified(g),
+	}
+}
+
+// Name is the unique engine name.
+func (e *LoginOffline) Name() string {
+	return "LoginOffline"
+}
+
+// Prereqs returns the engine prereqs.
+func (e *LoginOffline) Prereqs() Prereqs {
+	return Prereqs{}
+}
+
+// RequiredUIs returns the required UIs.
+func (e *LoginOffline) RequiredUIs() []libkb.UIKind {
+	return []libkb.UIKind{}
+}
+
+// SubConsumers returns the other UI consumers for this engine.
+func (e *LoginOffline) SubConsumers() []libkb.UIConsumer {
+	return nil
+}
+
+func (e *LoginOffline) Run(ctx *Context) error {
+	if err := e.run(ctx); err != nil {
+		return err
+	}
+
+	e.G().Log.Debug("LoginOffline success, sending login notification")
+	e.G().NotifyRouter.HandleLogin(string(e.G().Env.GetUsername()))
+	e.G().Log.Debug("LoginOffline success, calling login hooks")
+	e.G().CallLoginHooks()
+
+	return nil
+}
+
+func (e *LoginOffline) run(ctx *Context) error {
+	var gerr error
+	aerr := e.G().LoginState().Account(func(a *libkb.Account) {
+		var in bool
+		in, gerr = a.LoggedInProvisioned()
+		if gerr != nil {
+			e.G().Log.Debug("LoginOffline: LoggedInProvisioned error: %s", gerr)
+			return
+		}
+		if !in {
+			e.G().Log.Debug("LoginOffline: LoggedInProvisioned says not logged in")
+			gerr = libkb.LoginRequiredError{}
+			return
+		}
+
+		// current user has a valid session file
+
+		// check ActiveDevice cache
+		uid, deviceID, sigKey, encKey := e.G().ActiveDevice.AllFields()
+		if sigKey != nil && encKey != nil {
+			if uid.Equal(a.GetUID()) && deviceID.Eq(a.GetDeviceID()) {
+				// since they match, good to go
+				return
+			}
+		}
+
+		// nothing cached, so need to load the locked keys and unlock them
+		// with secret store
+
+		uid = e.G().Env.GetUID()
+		deviceID = e.G().Env.GetDeviceIDForUID(uid)
+
+		// use the UPAKLoader with StaleOK in order to get cached upak
+		arg := libkb.NewLoadUserByUIDArg(ctx.NetContext, e.G(), uid)
+		arg.PublicKeyOptional = true
+		arg.StaleOK = true
+		arg.LoginContext = a
+		upak, _, err := e.G().GetUPAKLoader().Load(arg)
+		if err != nil {
+			e.G().Log.Warning("upak.Load err: %s", err)
+			gerr = err
+			return
+		}
+
+		// find the sibkey
+		var sibkey *keybase1.PublicKey
+		for _, key := range upak.Base.DeviceKeys {
+			e.G().Log.Warning("device key: %+v", key)
+			if key.DeviceID.Eq(deviceID) && key.IsSibkey == true {
+				e.G().Log.Warning("device key match: %+v", key)
+				sibkey = &key
+				break
+			}
+		}
+		if sibkey == nil {
+			gerr = errors.New("no sibkey found")
+			return
+		}
+
+		// find the subkey
+		var subkey *keybase1.PublicKey
+		for _, key := range upak.Base.DeviceKeys {
+			if !key.IsSibkey && key.ParentID == sibkey.KID.String() {
```

Comment:
I can make it into one here, but assume there's a reason for it to be a string in the protocol?